src/python/zquantum/core/measurement_test.py

```diff
@@ -31,223 +30,271 @@
+    expected_bitstring = (1, 0, 0, 0)
+    amplitudes = np.array([0] * (2 ** n_qubits)).reshape(2 ** n_qubits, 1)
+    amplitudes[1] = 1  # |0001> will be measured in all cases.
+    wavefunction = Wavefunction(amplitudes)
+    sample = set(sample_from_wavefunction(wavefunction, 500))
+    assert len(sample) == 1
+    assert sample.pop() == expected_bitstring
+
+
+def test_sample_from_wavefunction_row_vector():
+    n_qubits = 4
+    # NOTE: our indexing places the state of qubit i at the ith index of the tuple. Hence |01> will result in
+    # the tuple (1, 0)
+    expected_bitstring = (1, 0, 0, 0)
+    amplitudes = np.array([0] * (2 ** n_qubits))
+    amplitudes[1] = 1  # |0001> will be measured in all cases.
+    wavefunction = Wavefunction(amplitudes)
+    sample = set(sample_from_wavefunction(wavefunction, 500))
+    assert len(sample) == 1
+    assert sample.pop() == expected_bitstring
+
+
+def test_sample_from_wavefunction_list():
+    n_qubits = 4
+    # NOTE: our indexing places the state of qubit i at the ith index of the tuple. Hence |01> will result in
+    # the tuple (1, 0)
+    expected_bitstring = (1, 0, 0, 0)
+    amplitudes = [0] * (2 ** n_qubits)
+    amplitudes[1] = 1  # |0001> will be measured in all cases.
+    wavefunction = Wavefunction(amplitudes)
+    sample = set(sample_from_wavefunction(wavefunction, 500))
+    assert len(sample) == 1
+    assert sample.pop() == expected_bitstring
+
+
+def test_parities_io():
+    measurements = [(1, 0), (1, 0), (0, 1), (0, 0)]
+    op = IsingOperator("[Z0] + [Z1] + [Z0 Z1]")
+    parities = get_parities_from_measurements(measurements, op)
+    save_parities(parities, "parities.json")
+    loaded_parities = load_parities("parities.json")
+    assert np.allclose(parities.values, loaded_parities.values)
+    assert len(parities.correlations) == len(loaded_parities.correlations)
+    for i in range(len(parities.correlations)):
+        assert np.allclose(parities.correlations[i], loaded_parities.correlations[i])
+    remove_file_if_exists("parities.json")
+
+
+def test_get_expectation_values_from_parities():
+    parities = Parities(values=np.array([[18, 50], [120, 113], [75, 26]]))
+    expectation_values = get_expectation_values_from_parities(parities)
+
+    assert len(expectation_values.values) == 3
+    assert np.isclose(expectation_values.values[0], -0.47058823529411764)
+    assert np.isclose(expectation_values.values[1], 0.030042918454935622)
+    assert np.isclose(expectation_values.values[2], 0.48514851485148514)
+
+    assert len(expectation_values.estimator_covariances) == 3
+    assert np.allclose(
+        expectation_values.estimator_covariances[0],
+        np.array([[0.014705882352941176]]),
+    )
+    assert np.allclose(
+        expectation_values.estimator_covariances[1], np.array([[0.00428797]])
+    )
+
+    assert np.allclose(
+        expectation_values.estimator_covariances[2], np.array([[0.0075706]])
+    )
+
+
+def test_expectation_vafmneaslues_to_real():
```

Comment:
Looks like this test name got mangled, should be `test_expectation_values_to_real `

Reply:
Whoops... I'm gonna blame that one on Summit 😆

Reply:
Addressed here: 
cedc8bb